test/bucket_id_test.lua

```diff
@@ -0,0 +1,227 @@
+    g.schema.spaces.space.sharding_func = {
+        body = helper.sharding_func_body
+    }
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local id, err = ddl.bucket_id('space', REFERENCE_KEY)
+    t.assert_equals(err, nil)
+    t.assert_equals(type(id), 'number')
+    t.assert_equals(id, REFERENCE_BUCKET_ID)
+end
+
+function g.test_bucket_id_sharding_func_user()
+    local sharding_func_name = 'sharding_func'
+    rawset(_G, sharding_func_name, helper.sharding_func)
+    g.schema.spaces.space.sharding_func = sharding_func_name
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local id, err = ddl.bucket_id('space', REFERENCE_KEY)
+    t.assert_equals(err, nil)
+    t.assert_equals(type(id), 'number')
+    t.assert_equals(id, REFERENCE_BUCKET_ID)
+end
+
+function g.test_bucket_id_sharding_func_table_call()
+    local sharding_func_name = 'sharding_func_table_call'
+    local tbl = {}
+    setmetatable(tbl, {
+        __call = helper.sharding_func
+    })
+    rawset(_G, sharding_func_name, tbl)
+    g.schema.spaces.space.sharding_func = sharding_func_name
+
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    local id, err = ddl.bucket_id('space', REFERENCE_KEY)
+    t.assert_equals(err, nil)
+    t.assert_equals(type(id), 'number')
+    t.assert_equals(id, REFERENCE_BUCKET_ID)
+end
+
+function g.test_bucket_id_sharding_func_userdata_call()
+    local sharding_func_name = 'sharding_func_userdata_call'
+
+    -- Create a userdata object.
+    local userdata_obj = newproxy(true)
+    local mt = getmetatable(userdata_obj)
+    mt.__call = helper.sharding_func
+    rawset(_G, sharding_func_name, userdata_obj)
+    t.assert_equals(type(userdata_obj), 'userdata')
+
+    -- Set sharding function as a __call method of userdata object.
+    rawset(_G, sharding_func_name, userdata_obj)
+    g.schema.spaces.space.sharding_func = sharding_func_name
+
+    -- Set DDL schema.
+    local ok, err = ddl.set_schema(g.schema)
+    t.assert_equals(err, nil)
+    t.assert_equals(ok, true)
+
+    -- Make sure __call method of userdata object can be used by bucket_id().
+    local id, err = ddl.bucket_id('space', REFERENCE_KEY)
+    t.assert_equals(err, nil)
+    t.assert_equals(type(id), 'number')
+    t.assert_equals(id, REFERENCE_BUCKET_ID)
+end
+
+function g.test_bucket_id_sharding_func_cdata_call()
+    local sharding_func_name = 'sharding_func_cdata_call'
+
+    -- Create a cdata object.
+    ffi.cdef[[
+       typedef struct {
+	    double x;
+	    double y;
+       } point_t;
+    ]]
+    local point = ffi.metatype('point_t', {
+        __call = helper.sharding_func
+    })
+    t.assert_equals(type(point), 'cdata')
+
+    -- Set sharding function as a __call method of cdata object.
+    rawset(_G, sharding_func_name, point())
```

Comment:
1. The comment does not look relevant to the code (we don't set a metatype with __call here).
2. Strange. The cdata object does not participate in the test at all (only `point()` result, but `point` should be here). I don't know what is going on here, but it seems that the test should fail on `ddl.set_schema()`.

Reply:
1. Fixed.
2. Indeed test for cdata was incorrect.

```lua
--- a/test/bucket_id_test.lua
+++ b/test/bucket_id_test.lua
@@ -200,12 +200,13 @@ function g.test_bucket_id_sharding_func_cdata_call()
            double y;
        } point_t;
     ]]
-    local point = ffi.metatype('point_t', {
+    ffi.metatype('point_t', {
         __call = helper.sharding_func
     })
+    local point = ffi.new('point_t')
     t.assert_equals(type(point), 'cdata')
 
-    rawset(_G, sharding_func_name, point())
+    rawset(_G, sharding_func_name, point)
     g.schema.spaces.space.sharding_func = sharding_func_name
 
     -- Set DDL schema.
```